From: Mauro Carvalho Chehab <mchehab@redhat.com>
Date: Fri, 2 Jul 2010 23:15:32 -0400
Subject: [x86_64] mce: fix misplaced `continue' in mce.c
Message-id: <20100702201532.7a142f9b@pedra>
Patchwork-id: 26703
O-Subject: [RHEL5 BZ#468877 3/5] x86_64: fix misplaced `continue' in mce.c
Bugzilla: 468877
RH-Acked-by: Jarod Wilson <jarod@redhat.com>

Patch-mainline: commit 4f84e4be53a04a65d97bf0faa0c8f99e29bc0170
Original patch From: Joshua Wise <jwise@google.com>
Date:   Sat Jun 23 17:16:45 2007 -0700
Acked-by: ak@suse.de
References: 344239
Subject: x86_64: fix misplaced `continue' in mce.c

......
    Background:
      When a userspace application wants to know about machine check events, it
      opens /dev/mcelog and does a read(). Usually, we found that this interface
      works well, but in some cases, when the system was taking large numbers of
      machine check exceptions, the read() would hang. The system would output a
      soft-lockup warning, and the daemon reading from /dev/mcelog would suck up
      as much of a single CPU as it could spinning in system space.

    Description:
      This patch fixes this bug. In particular, there was a "continue" inside a
      timeout loop that presumably was intended to break out of the outer loop,
      but instead caused the inner loop to continue. This patch also makes the
      condition for the break-out a little more evident by changing a
      !time_before to a time_after_eq.

    Result:
      The read() no longer hangs in this test case.

    Testing:
      On my system, I could replicate the bug with the following command:
        # for i in `seq 15000`; do ./inject_sbe.sh; done
      where inject_sbe.sh contains commands to inject a single-bit error into the
      next memory write transaction.

    Patch:
      This patch is against git f1518a088bde6aea49e7c472ed6ab96178fcba3e.

    Signed-off-by: Joshua Wise <jwise@google.com>
    Signed-off-by: Tim Hockin <thockin@google.com>
    Cc: Andi Kleen <ak@suse.de>
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
Signed-off-by: Mauro Carvalho Chehab <mchehab@redhat.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86_64/kernel/mce.c b/arch/x86_64/kernel/mce.c
index 3f223e9..510f95b 100644
--- a/arch/x86_64/kernel/mce.c
+++ b/arch/x86_64/kernel/mce.c
@@ -490,15 +490,17 @@ static ssize_t mce_read(struct file *filp, char __user *ubuf, size_t usize, loff
 	for (i = 0; i < next; i++) {		
 		unsigned long start = jiffies;
 		while (!mcelog.entry[i].finished) {
-			if (!time_before(jiffies, start + 2)) {
+			if (time_after_eq(jiffies, start + 2)) {
 				memset(mcelog.entry + i,0, sizeof(struct mce));
-				continue;
+				goto timeout;
 			}
 			cpu_relax();
 		}
 		smp_rmb();
 		err |= copy_to_user(buf, mcelog.entry + i, sizeof(struct mce));
 		buf += sizeof(struct mce); 
+ timeout:
+		;
 	} 
 
 	memset(mcelog.entry, 0, next * sizeof(struct mce));
